Refactor blockchaininterface: fee estimation code

kristapsk opened the pull request:
Part of splitting #1462 into smaller PRs for easier reviewing and testing.

PulpCattel commented on jmclient/jmclient/blockchaininterface.py:
```diff
@@ -447,29 +481,10 @@ def estimate_fee_per_kb(self, N):
-                msg = msg + " (randomized for privacy)"
-            log.info(msg + ": " + btc.fee_per_kb_to_str(retval))
-            return int(retval)
+                return btc.btc_to_sat(estimate)
+            else:  # cannot get a valid estimate after `tries` tries:
+                log.warn("Could not source a fee estimate from Core")
+                return None
```
in 68352d047a636496ec5272c41b4d168ecf481c40:

This `else` will stop the loop, which I don't think it is what we want (as per the comment `cannot get a valid estimate after tries tries`)

I believe in the old code this `else` was used like [this](https://stackoverflow.com/questions/9979970/why-does-python-use-else-after-for-and-while-loops).

kristapsk replied:
Identation issue, fixed in 3a8f456bd9d4e7297cb502224757a2537fc1ffb7.

PulpCattel replied:
I'm not sure how the `else` works in this case (with `return` inside the loop instead of `break`), but in any case I think we don't need it anymore, right?

We can do just:

```python3
    def _estimate_fee_basic(self, tx_fees: int) -> Optional[int]:
        # Special bitcoin core case: sometimes the highest priority
        # cannot be estimated in that case the 2nd highest priority
        # should be used instead of falling back to hardcoded values
        tries = 2 if tx_fees == 1 else 1
        for i in range(tries):
            rpc_result = self._rpc('estimatesmartfee', [tx_fees + i])
            if not rpc_result:
                # in case of connection error:
                return None
            estimate = rpc_result.get('feerate')
            # `estimatesmartfee` will currently return in the format
            # `{'errors': ['Insufficient data or no feerate found'], 'blocks': N}`
            # if it is not able to make an estimate. We insist that
            # the 'feerate' key is found and contains a positive value:
            if estimate and estimate > 0:
                return btc.btc_to_sat(estimate)
        # cannot get a valid estimate after `tries` tries:
        log.warn("Could not source a fee estimate from Core")
```

kristapsk replied:
Yes, that `else` is actually not needed anymore, code is easier to read without it. Addressed in 36929c0453aa88dd3257162f1462c36f4ddd785b.

PulpCattel commented on jmclient/jmclient/blockchaininterface.py:
```diff
@@ -665,7 +680,7 @@ def __init__(self, jsonRpc, wallet_name):
         self.destn_addr = self._rpc("getnewaddress", [])
 
-    def estimate_fee_per_kb(self, N):
+    def estimate_fee_per_kb(self, N: int) -> int:
```
nit in 68352d047a636496ec5272c41b4d168ecf481c40:

Same thing about `N` variable name.

PulpCattel commented on jmclient/test/commontest.py:
```diff
@@ -131,7 +135,7 @@ def query_utxo_set(self, txouts, includeconfs=False):
         return result
 
-    def estimate_fee_per_kb(self, N):
+    def estimate_fee_per_kb(self, N: int) -> int:
```
nit in 68352d047a636496ec5272c41b4d168ecf481c40:

Same thing about `N` variable name.

PulpCattel reviewed (commented):
This PR refactors the `estimate_fee_per_kb` method. Mostly splitting it into multiple methods, adding the relevant abstract methods, and adding type hint.
Tests pass.

I think it should be possible to clean up the `estimate_fee_per_kb` method further. In particular, the difference between the two cases (block target and sat/vb) seems very small and maybe we could de-duplicate it further.

kristapsk commented on jmclient/jmclient/blockchaininterface.py:
```diff
@@ -63,15 +56,90 @@ def import_addresses_if_needed(self, addresses, wallet_name):
+        or None in case of error.
+        """
+
+    def _fee_per_kb_has_been_manually_set(self, N: int) -> bool:
+        """if the 'block' target is higher than 1000, interpret it
```
`conf_target` would mean that it is always confirmation target in number of blocks, but it isn't for values above 1000, then it's sat/kvB.

PulpCattel replied:
Yes, that's true. While certainly not perfect, I think a simple way we could look at it is that this distinction is made inside the method itself. I.e., We are passing to it a conf target of, e.g., 9000, and the method will make the distinction. But we still passed a conf target to it.
The docstring also mentions a block target but we only have `N`.

If we want to be fancy we could use something like "conf_target_or_feerate" instead. I personally don't mind the verbose names.

Or what do you think of just `target`? Might be a decent middle ground to mean both a block and feerate target?

kristapsk replied:
Maybe `tx_fees`, which is how it's called in `joinmarket.cfg'?

PulpCattel replied:
I'm okay with it.

kristapsk commented on jmclient/jmclient/blockchaininterface.py:
```diff
@@ -63,15 +56,90 @@ def import_addresses_if_needed(self, addresses, wallet_name):
+        how this is distinguished).
+        In both cases it is prevented from falling below the current
+        minimum feerate for tx to be accepted into node's mempool.
+        In case of failure to connect, None is returned.
```
Addressed in 989a221f3a3dd619cfee17b124f8c464d9592026

kristapsk commented on jmclient/jmclient/blockchaininterface.py:
```diff
@@ -63,15 +56,90 @@ def import_addresses_if_needed(self, addresses, wallet_name):
+            if N_res < mempoolminfee_in_sat:
+                log.info("Using this mempool min fee as tx feerate: " +
+                    btc.fee_per_kb_to_str(mempoolminfee_in_sat) + ".")
+                return int(mempoolminfee_in_sat_randomized)
```
Addressed in aa9d50e8ced04bacc7e33e007b6f859e61e44752

kristapsk commented on jmclient/jmclient/blockchaininterface.py:
```diff
@@ -63,15 +56,90 @@ def import_addresses_if_needed(self, addresses, wallet_name):
+        if retval < mempoolminfee_in_sat:
+            log.info("Using this mempool min fee as tx feerate: " +
+                btc.fee_per_kb_to_str(mempoolminfee_in_sat) + ".")
+            return int(mempoolminfee_in_sat_randomized)
```
Addressed in aa9d50e8ced04bacc7e33e007b6f859e61e44752

kristapsk commented:
@PulpCattel I think I have addressed all the review comments, if this looks good, would like to squash and merge soon.

> I think it should be possible to clean up the `estimate_fee_per_kb` method further. In particular, the difference between the two cases (block target and sat/vb) seems very small and maybe we could de-duplicate it further.

This could go as a follow up PR in future.

PulpCattel commented on jmclient/jmclient/blockchaininterface.py:
```diff
@@ -63,15 +56,90 @@ def import_addresses_if_needed(self, addresses, wallet_name):
+            retval = fallback_fee
+        else:
+            retval = random.uniform(retval,
+                retval * float(1 + tx_fees_factor))
```
Can you clarify this? It still happens in 936695aa424b8b7d905cf91441c1ce68118a9190.

kristapsk replied:
Forgot about this. Yes, IMO it makes sense to randomize fallback fee too. Addressed in 31d0692b258805466a66fffa20ac4178bc39a94d.

PulpCattel commented on jmclient/jmclient/blockchaininterface.py:
```diff
@@ -390,52 +458,18 @@ def get_unspent_indices(self, transaction):
-                log.info(msg + ": " + btc.fee_per_kb_to_str(N_res) + ".")
-                return int(N_res)
-
+    def _estimate_fee_basic(self, N: int) -> Optional[int]:
```
In 936695aa424b8b7d905cf91441c1ce68118a9190:

This should be done I think. Here we are always passing a `conf_target`. If we want to use `tx_fees` here too, we should change the declaration at line 66 too.

kristapsk replied:
Correct, fixed in e82cfaf735604658011df5318882e6c14ee9818a.

kristapsk commented:
@PulpCattel Looks ok now?

PulpCattel commented:
Yes, tested 36929c0453aa88dd3257162f1462c36f4ddd785b. Tests pass, and I've done some basic manual testing on signet.

kristapsk commented on jmclient/jmclient/blockchaininterface.py:
```diff
@@ -63,14 +56,96 @@ def import_addresses_if_needed(self, addresses, wallet_name):
+            mempoolminfee_in_sat, mempoolminfee_in_sat * float(1 + tx_fees_factor))
+
+        if self._fee_per_kb_has_been_manually_set(tx_fees):
+            N_res = random.uniform(tx_fees, tx_fees * float(1 - tx_fees_factor))
```
Here's typo bug (found by manual testing), should be `+` instead of `-`, will update soon.

PulpCattel replied:
Ah right, we randomize only upwards, so a lower value is a bug. Good catch.

This is one of the reasons we should eventually de-duplicate more out of this function.

kristapsk replied:
Agree about de-duplication and also fee estimation code currently lacks any unit tests, that could have catched this.

kristapsk commented:
CI test failure is known unrelated issue that happens from time to time, merging.
